test/studies/shootout/binary-trees/binarytrees-pool.chpl

```diff
@@ -8,31 +8,35 @@
 
 use Allocators;
 
-config const n = 10,         // the maximum tree depth
-             globalPoolSize = (2 ** 31)-1,
-             localPoolSize = globalPoolSize;
-
-var globalPool = new bumpPtrMemPool(globalPoolSize, alignment=0);
+config const n = 10;                       // the maximum tree depth
 
 proc main() {
   const minDepth = 4,                      // the shallowest tree
         maxDepth = max(minDepth + 2, n),   // the deepest normal tree
         strDepth = maxDepth + 1,           // the depth of the "stretch" tree
-        depths = minDepth..maxDepth by 2;  // the range of depths to create
+        depths = minDepth..maxDepth by 2,  // the range of depths to create
+        nodeSize = 24;                     // the approximate size of a node
   var stats: [depths] (int,int);           // stores statistics for the trees
+  const poolSize =
+    (((2**(maxDepth+2))-1)*nodeSize)*3;    // how much each pool allocates
+                                           // based on the deepest tree,
+                                           // with a factor of 3 for safety
```

Comment:
The +2 is required because the largest tree allocated has a depth of `maxDepth+1`, and we need +1 for the formula `2**(d+1)-1`. And experimentally, I can show its required because the allocator is bounds checked and runs out of memory (or segfaults without `--checks`)

Sadly, the fudge factor is needed for a few reasons, and I actually realized the fudge factor I put is wrong....but only for some machines. TL;DR the global fudge factor can be "removed" provided we used two global pools and that each local pool uses a local fudge factor.

To start, the "globalPool" is used to allocate two trees, `strTree` and `llTree`. So the pool needs to be big enough to allocate both trees. I think technically, `strTree` should be deallocated at the end of its scope (its "short-lived"). In the most recent commit, I have resolved this by using two `pools`, one for the short-lived tree that is deallocated immediately and one for the long lived tree that is deallocated at the end of `main`. I checked the performance on my machine, and I don't think this one extra allocation/deallocation will hurt performance. Making this change is arguably more correct and removes the need for the fudge factor.

Second, the fudge factor is needed because each "localPool" allocates many trees of max depth. So without the fudge factor, they may run out memory if there are not enough tasks to split up the trees. I realized the current factor of 3 is not enough when there is only 1 task (i.e. `CHPL_RT_NUM_THREADS_PER_LOCALE=1 ./binarytrees-pool --n=21` would run out of memory, but `CHPL_RT_NUM_THREADS_PER_LOCALE=8 ./binarytrees-pool --n=21` would not). On my machine I tested more thoroughly and found that with 1 task a fudge factor of 8 was sufficient. This is also in the most recent commit.

Reply:
> The +2 is required because the largest tree allocated has a depth of maxDepth+1

You're right of course.  I forgot that there was a one-larger tree, and also was misreading the output of the program, thinking the first tree printed was of maxDepth, when it's the maxDepth+1 case—so I thought it was validating my math when it wasn't.

And thanks for the fudge factor explanation.  I was forgetting that these were task-private rather than iteration-private.  And we obviously wouldn't want to make it iteration-private.  But on your first point:

> Second, the fudge factor is needed because each "localPool" allocates many trees of max depth. So without the fudge factor, they may run out memory if there are not enough tasks to split up the trees.

Isn't each of these trees also de-allocated before the next one is allocated?  Oh, no, because they're unmanaged rather than owned so deallocated when the pool leaves scope.  OK, got it.

Does it seem correct that we could make the 8 less magic by computing the local pool size as something like `iterations/here.maxTaskPar * 2**(depth+1)` or something like that?  (giving it a nice name along the way).  But is the precision and memory savings worth the extra characters…?  I suppose the main advantage would be that if he went to a 16-core or 32-core or 128-core node in the future, it could significantly reduce our memory usage?  (Is it right that we're allocating about a Gig per task in the performance problem size?)

Do you know:  What units is his `mem` column on the website measured in?  And how will this version compare against our current "477,824" high-water mark?

Reply:
The formula `iterations/here.maxTaskPar * 2**(depth+1)` seems to work fine with 1 thread on my laptop 🎉 

I have no idea what the units of that memory usage are, but based on the sites docs they use maximum resident set size as measured by python `waitpid`.

So using this python code, I measured `binarytrees-inner` and `binarytrees-pool`
```python
import sys
import subprocess
import os
args = sys.argv[1:]
p = subprocess.Popen(args)
ru = os.wait4(p.pid, 0)[2]
print("Maximum rss {} GB".format(ru.ru_maxrss/1024/1024))
```

On my laptop, I got 726 GB for `inner` and 1522 GB for `pool`. I am not sure I totally believe those numbers (a thousands GBs seems strange), but I think the relative comparison of "twice as much memory" is useful.

Reply:
Also I tried to make the local `pool` a `const`, but apparently that doesn't work: https://github.com/chapel-lang/chapel/issues/26017